internal/cache/file/cache_handle.go

```diff
@@ -92,12 +93,50 @@ func (fch *CacheHandle) shouldReadFromCache(jobStatus *downloader.JobStatus, req
 	return err
 }
 
+// checkEntryInFileInfoCache checks if entry is present for a given object in
+// file info cache with same generation and offset at least equal to
+// requiredOffset. It returns nil if entry is present, otherwise returns an error
+// with appropriate message.
+func (fch *CacheHandle) checkEntryInFileInfoCache(bucket gcs.Bucket, object *gcs.MinObject, requiredOffset int64) error {
+	fileInfoKey := data.FileInfoKey{
+		BucketName: bucket.Name(),
+		ObjectName: object.Name,
+	}
+	fileInfoKeyName, err := fileInfoKey.Key()
+	if err != nil {
+		return fmt.Errorf("read: while creating key: %v", fileInfoKeyName)
+	}
+	// Look up of file being read in file info cache is required to update the LRU
+	// order on every read request from kernel i.e. with every read request from
+	// kernel, the file being read becomes most recently used.
+	fileInfo := fch.fileInfoCache.LookUp(fileInfoKeyName)
+
+	// The offset and generation checks below are required because it may happen
+	// that file being read is evicted from cache during or after reading the
+	// required offset from local cached file to `dst` buffer in previous code
+	// block.
+	if fileInfo == nil {
+		err = fmt.Errorf("%v: no entry found in file info cache for key %v", util.InvalidFileInfoCacheErrMsg, fileInfoKeyName)
+		return err
+	}
+	fileInfoData := fileInfo.(data.FileInfo)
+	if requiredOffset > int64(fileInfoData.Offset) {
```

Comment:
We call this method after reading from the file in cache. Now, if this condition is true that means corresponding job and fileInfoEntry is not consistent. In this case, we should close the fileHandle and then fallback to GCS. For that we need to return different error other than FallBackToGCS.

As randomReader doesn't close cacheHandle for FallBackToGCSErrMsg.

Reply:
As discussed offline, we don't need comparison of offsets as we are already comparing the number of bytes read with numBytesRequested in previous code.